processor/src/test/resources/fixtures/org/mapstruct/ap/test/bugs/_1685/UserMapperImpl.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1685;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.annotation.Generated;
+
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor",
+    date = "2019-01-27T12:40:32+0100",
+    comments = "version: , compiler: Eclipse JDT (Batch) 1.2.100.v20160418-1457, environment: Java 1.8.0_181 (Oracle Corporation)"
+)
+public class UserMapperImpl implements UserMapper {
+
+    @Override
+    public UserDTO userToUserDTO(User user) {
+        if ( user == null ) {
+            return null;
+        }
+
+        UserDTO userDTO = new UserDTO();
+
+        userDTO.setContactDataDTO( userToContactDataDTO( user ) );
+        userDTO.setName( user.getName() );
+
+        return userDTO;
+    }
+
+    @Override
+    public void updateUserFromUserDTO(UserDTO userDTO, User user) {
+        if ( userDTO == null ) {
+            return;
+        }
+
+        String[] settings1 = userDTOContactDataDTOSettings( userDTO );
+        if ( settings1 != null ) {
+            user.setSettings( Arrays.copyOf( settings1, settings1.length ) );
+        }
+        else {
+            user.setSettings( null );
+        }
+        if ( userDTOContactDataDTOPreferences( userDTO ) != null ) {
```

Comment:
I checked it.. But it is existing functionality. It cannot be changed that easily. The point is that on low level, the source reference is either: 
1. a local var
2. original source reference

if 1. is defined, it will take 1. in stead of the original  source reference.

The iterator variable occupies the position of local var. But we need one extra, to capture the result of the nested method (which is set to the local var).. Still following me? 

Well, point being, the `CommonMacros.ftl` needs to be refactored.. I could include a new precedence rule:
1. loopVar
2. localVar
3. sourceReference 
(and have better names). 

But is it really part of this PR? or shall I open a new issue.. As far as I'm concerned its optimisation of code that already showed that behaviour. Should we do this in 1.4 / 1.3.1?

Reply:
Checkout the `SourceReference.tftl` template.. then my story above makes sense..

Reply:
I think it can belong to a new PR, and you are saying that before it was like that as well? In that case I think we are OK.

Reply:
see #1700